File: 프로그래머스/3/250134. ［PCCP 기출문제］ 4번 ／ 수레 움직이기/［PCCP 기출문제］ 4번 ／ 수레 움직이기.cpp
#include <string>
#include <vector>
#include <climits>
using namespace std;

struct xy{
    int x;
    int y;
};
int Min = INT_MAX;
int n,m;

int visit[5][5][2];
int map[5][5];
int dx[] = {1,-1,0,0};
int dy[]  = {0,0,1,-1};
void dfs(xy R, xy B, int sta,int cnt,int turn, int sucR, int sucB){
    if(R.x < 0||R.x >= n || R.y < 0 || R.y >= m) return;
    if(B.x < 0||B.x >= n || B.y < 0 || B.y >= m) return;
    if(R.x == B.x and R.y == B.y) return;
    
    int nx = (sta == 0 ? R.x : B.x);        //이전에 0을 움직임 = 빨강
    int ny = (sta == 0 ? R.y : B.y);
    
    if(map[nx][ny] == 5) return;
    if(visit[nx][ny][sta] == 1) return;
    
    if(sta == 1) //파란검사
        if(map[nx][ny] == 4) sucB = 1;
    
    if(sta == 0) //빨간 검사
        if(map[nx][ny] == 3) sucR = 1;
    
    if(cnt%2 == 1) turn++;
    if(sucR == 1 and sucB == 1){
        Min = min(Min,turn); return;
    }
    
    
    visit[nx][ny][sta] = 1;
    
    if(sta == 1){//파랑움직임, 빨강차례
        if(sucR == 1){//빨이미 되있으면 파랑움직임
            for(int i = 0; i<4; i++)
                dfs(R, {B.x + dx[i], B.y + dy[i]}, 1,0,turn + 1, sucR,sucB);//파랑움직임
        }
        else{
            for(int i = 0; i<4; i++)
                dfs({R.x + dx[i], R.y + dy[i]}, B, 0,cnt + 1,turn, sucR,sucB);
        }

    }
    else{
        if(sucB == 1){
            for(int i = 0; i<4; i++)
                 dfs({R.x + dx[i], R.y + dy[i]}, B, 0,0,turn + 1, sucR,sucB);
        }
        else{
            for(int i = 0; i<4; i++)
                dfs(R, {B.x + dx[i], B.y + dy[i]}, 1,cnt + 1,turn, sucR,sucB);
        }

    }
    visit[nx][ny][sta] = 0;
    
}

int solution(vector<vector<int>> maze) {
    int answer = 0;
    
    xy RT,BT,RS,BS;
    n = maze.size();
    m = maze[0].size();
    for(int i = 0; i< maze.size(); i++){
        for(int j = 0; j<maze[0].size(); j++){
            if(maze[i][j] == 1) RS = {i,j};
            else if(maze[i][j] == 2) BS = {i,j};
            else if(maze[i][j] == 3) RT = {i,j};
            else if(maze[i][j] == 4) BT = {i,j};
            map[i][j] = maze[i][j];
        }
    }
    
    dfs(RS,BS,0,0,0,0,0);
    
    for(int i = 0; i<5;i++)
        for(int j = 0; j<5;j++)
            for(int k = 0 ;k<2;k++)
                visit[i][j][k] = 0;
    
    
    dfs(RS,BS,1,0,0,0,0);
    answer = (Min != INT_MAX ? Min : 0);
    return answer;
}

File: 프로그래머스/3/250134. ［PCCP 기출문제］ 4번 ／ 수레 움직이기/［PCCP 기출문제］ 4번 ／ 수레 움직이기.java
import java.io.*;
import java.util.*;

class Solution {
    static int[][] map;
    static int[][][] visit;
    static int n, m, an;
    static int[] dx = {1,-1,0,0};
    static int[] dy = {0,0,1,-1};
    
    public int solution(int[][] maze) {
        int answer = 0;
        
        map = maze;
        n = maze.length;
        m = maze[0].length;
        an = Integer.MAX_VALUE;
        visit = new int[n][m][2];
        
        xy R = new xy(0 ,0);
        xy B = new xy(0, 0);
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(map[i][j]==1){
                    R.x = i; R.y = j;
                }
                else if(map[i][j] == 2){
                    B.x = i; B.y = j;
                }
            }
        }
        
        dfs(R, B, 0, 0, 0, 0);
        
        answer = an;
        if(answer == Integer.MAX_VALUE) answer = 0;
            
        return answer;
    }
    
    static void dfs(xy R, xy B,int turn,int sucR,int sucB, int cnt){
        if(R.x < 0 || R.x >= n || R.y < 0 || R.y >= m) return;
        if(B.x < 0 || B.x >= n || B.y < 0 || B.y >= m) return;
        if(R.x == B.x && R.y == B.y ) return;
        if(map[R.x][R.y] == 5 || map[B.x][B.y] == 5) return;
        if(turn == 0 && visit[R.x][R.y][turn] == 1) return;
        if(turn == 1 && visit[B.x][B.y][turn] == 1) return;
        
        if(map[R.x][R.y] == 3) sucR = 1;
        if(map[B.x][B.y] == 4) sucB = 1;
        if(sucR == 1 && sucB == 1){
            an = Math.min(an, cnt);
            return;
        }
        
        xy ball;
        if(turn == 0) ball = new xy(R.x, R.y);
        else ball = new xy(B.x, B.y);
        
        visit[ball.x][ball.y][turn] = 1;
        
        if(sucB == 1){
            for(int i = 0; i < 4; i++)
                dfs(new xy(R.x + dx[i], R.y + dy[i]), B, 0, sucR, sucB, cnt + 1);
        }
        else if(sucR == 1){
            for(int i = 0; i < 4; i++)
                dfs(R, new xy( B.x + dx[i], B.y + dy[i] ), 1, sucR, sucB, cnt + 1);
        }
        
        else if(turn == 0){
            for(int i = 0; i < 4; i++)
                dfs(R, new xy( B.x + dx[i], B.y + dy[i] ), 1, sucR, sucB, cnt);
        }
        else if(turn == 1){
            for(int i = 0; i < 4; i++)
                dfs(new xy(R.x + dx[i], R.y + dy[i]), B, 0, sucR, sucB, cnt + 1);
        }
        visit[ball.x][ball.y][turn] = 0;
    }
    
    static class xy{
        int x, y;
        xy(int x, int y){
            this.x = x;
            this.y = y;
        }
    }
}